Cgen unwraps unchecked option type

### Describe the bug

<title>

### Reproduction Steps

```v
@[heap]
struct Chan[T] {
	c chan T
}

pub type EventListener[T] = fn (T) !
pub type Check[T] = fn (T) bool

struct EventWaiter[T] {
	check ?Check[T]
	c     Chan[T]
}

pub struct EventController[T] {
mut:
	id        int
	wait_fors map[int]EventWaiter[T]
	listeners map[int]EventListener[T]
}

fn (mut ec EventController[T]) generate_id() int {
	return ec.id++
}

pub fn (mut ec EventController[T]) emit(e T) {
	for i, w in ec.wait_fors {
		mut b := false
		c := w.check
		if c != none {
			b = c(e)
		} else {
			b = true
		}
		println('${i} passed: ${b}')
	}
}

struct Foo {}
struct Bar {}

fn use[T](_ EventController[T]) {}

fn main() {
	use(EventController[Foo]{})
	use(EventController[Bar]{})
}
```


### Expected Behavior

Successful compile

### Current Behavior

```c
==================
C:/Users/mclr/AppData/Local/Temp/v_0/test.5224352848663389627.tmp.c:18507: warning: assignment from incompatible pointer type
C:/Users/mclr/AppData/Local/Temp/v_0/test.5224352848663389627.tmp.c:18536: error: cannot convert 'struct _option_anon_fn_main__foo__bool' to 'unsigned char (*)(struct main__Foo)'
...
==================
(Use `v -cg` to print the entire error message)

builder error:
==================
C error. This should never happen.

This is a compiler bug, please report it using `v bug file.v`.

https://github.com/vlang/v/issues/new/choose

You can also use #help on Discord: https://discord.gg/vlang
```

### Possible Solution

_No response_

### Additional Information/Context

_No response_

### V version

V 0.4.3 cb28144

### Environment details (OS name and version, etc.)

```
V full version: V 0.4.3 57a33c3.cb28144
OS: windows, Microsoft Windows 11 Pro v22000 64-bit
Processor: 12 cpus, 64bit, little endian, 

getwd: D:\Games\Proekti\V\interactions
vexe: D:\Games\Proekti\V\v\v.exe
vexe mtime: 2024-01-01 12:03:01

vroot: OK, value: D:\Games\Proekti\V\v
VMODULES: OK, value: C:\Users\mclr\.vmodules
VTMP: OK, value: C:\Users\mclr\AppData\Local\Temp\v_0

Git version: git version 2.37.0.windows.1
Git vroot status: weekly.2023.52-38-gcb28144e
.git/config present: true

CC version:
thirdparty/tcc status: thirdparty-windows-amd64 a39eb79b
```
> [!NOTE]
> You can use the 👍 reaction to increase the issue's priority for developers.
>
> Please note that only the 👍 reaction to the issue itself counts as a vote.
> Other reactions and those to comments will not be taken into account.

--- a/vlib/v/gen/c/assign.v
+++ b/vlib/v/gen/c/assign.v
@@ -531,16 +531,13 @@ fn (mut g Gen) assign_stmt(node_ ast.AssignStmt) {
 				if is_inside_ternary {
 					g.out.write_string(util.tabs(g.indent - g.inside_ternary))
 				}
-				ret_styp := if val is ast.SelectorExpr {
-					g.typ(g.unwrap_generic(val_type))
-				} else {
-					g.typ(right_sym.info.func.return_type)
-				}
 				fn_name := c_fn_name(g.get_ternary_name(ident.name))
 
-				if val_type.has_flag(.option) {
+				if val_type.has_flag(.option) && val is ast.SelectorExpr {
+					ret_styp := g.typ(g.unwrap_generic(val_type))
 					g.write('${ret_styp} ${fn_name}')
 				} else {
+					ret_styp := g.typ(right_sym.info.func.return_type)
 					mut call_conv := ''
 					mut msvc_call_conv := ''
 					for attr in right_sym.info.func.attrs {

--- a/vlib/v/gen/c/assign.v
+++ b/vlib/v/gen/c/assign.v
@@ -531,34 +531,41 @@ fn (mut g Gen) assign_stmt(node_ ast.AssignStmt) {
 				if is_inside_ternary {
 					g.out.write_string(util.tabs(g.indent - g.inside_ternary))
 				}
-				ret_styp := g.typ(right_sym.info.func.return_type)
-
-				mut call_conv := ''
-				mut msvc_call_conv := ''
-				for attr in right_sym.info.func.attrs {
-					match attr.name {
-						'callconv' {
-							if g.is_cc_msvc {
-								msvc_call_conv = '__${attr.arg} '
-							} else {
-								call_conv = '${attr.arg}'
-							}
-						}
-						else {}
-					}
-				}
-				call_conv_attribute_suffix := if call_conv.len != 0 {
-					'__attribute__((${call_conv}))'
+				ret_styp := if val is ast.SelectorExpr {
+					g.typ(g.unwrap_generic(val_type))
 				} else {
-					''
+					g.typ(right_sym.info.func.return_type)
 				}
-
 				fn_name := c_fn_name(g.get_ternary_name(ident.name))
-				g.write('${ret_styp} (${msvc_call_conv}*${fn_name}) (')
-				def_pos := g.definitions.len
-				g.fn_decl_params(right_sym.info.func.params, unsafe { nil }, false)
-				g.definitions.go_back(g.definitions.len - def_pos)
-				g.write(')${call_conv_attribute_suffix}')
+
+				if val_type.has_flag(.option) {
+					g.write('${ret_styp} ${fn_name}')
+				} else {
+					mut call_conv := ''
+					mut msvc_call_conv := ''
+					for attr in right_sym.info.func.attrs {
+						match attr.name {
+							'callconv' {
+								if g.is_cc_msvc {
+									msvc_call_conv = '__${attr.arg} '
+								} else {
+									call_conv = '${attr.arg}'
+								}
+							}
+							else {}
+						}
+					}
+					call_conv_attribute_suffix := if call_conv.len != 0 {
+						'__attribute__((${call_conv}))'
+					} else {
+						''
+					}
+					g.write('${ret_styp} (${msvc_call_conv}*${fn_name}) (')
+					def_pos := g.definitions.len
+					g.fn_decl_params(right_sym.info.func.params, unsafe { nil }, false)
+					g.definitions.go_back(g.definitions.len - def_pos)
+					g.write(')${call_conv_attribute_suffix}')
+				}
 			} else {
 				if is_decl {
 					if is_inside_ternary {

--- a/vlib/v/gen/c/fn.v
+++ b/vlib/v/gen/c/fn.v
@@ -1884,12 +1884,9 @@ fn (mut g Gen) fn_call(node ast.CallExpr) {
 								if cast_sym.info is ast.Aggregate {
 									sym := g.table.sym(cast_sym.info.types[g.aggregate_type_idx])
 									g.write('${dot}_${sym.cname}')
-								} else if cast_sym.kind == .function {
-									if obj.orig_type.has_flag(.option) {
-										g.write('.data')
-									} else {
-										g.write('${dot}_${cast_sym.cname}')
-									}
+								} else if cast_sym.kind == .function
+									&& obj.orig_type.has_flag(.option) {
+									g.write('.data')
 								} else {
 									g.write('${dot}_${cast_sym.cname}')
 								}

--- a/vlib/v/gen/c/fn.v
+++ b/vlib/v/gen/c/fn.v
@@ -1860,8 +1860,12 @@ fn (mut g Gen) fn_call(node ast.CallExpr) {
 						}
 						if obj.smartcasts.len > 0 && is_cast_needed {
 							for typ in obj.smartcasts {
-								sym := g.table.sym(typ)
-								g.write('(*(${sym.cname})(')
+								sym := g.table.sym(g.unwrap_generic(typ))
+								if obj.orig_type.has_flag(.option) && sym.kind == .function {
+									g.write('(*(${sym.cname}*)(')
+								} else {
+									g.write('(*(${sym.cname})(')
+								}
 							}
 							for i, typ in obj.smartcasts {
 								cast_sym := g.table.sym(g.unwrap_generic(typ))
@@ -1880,6 +1884,10 @@ fn (mut g Gen) fn_call(node ast.CallExpr) {
 								if cast_sym.info is ast.Aggregate {
 									sym := g.table.sym(cast_sym.info.types[g.aggregate_type_idx])
 									g.write('${dot}_${sym.cname}')
+								} else if cast_sym.kind == .function {
+									if obj.orig_type.has_flag(.option) {
+										g.write('.data')
+									}
 								} else {
 									g.write('${dot}_${cast_sym.cname}')
 								}

--- a/vlib/v/gen/c/fn.v
+++ b/vlib/v/gen/c/fn.v
@@ -1887,6 +1887,8 @@ fn (mut g Gen) fn_call(node ast.CallExpr) {
 								} else if cast_sym.kind == .function {
 									if obj.orig_type.has_flag(.option) {
 										g.write('.data')
+									} else {
+										g.write('${dot}_${cast_sym.cname}')
 									}
 								} else {
 									g.write('${dot}_${cast_sym.cname}')

--- a/vlib/v/tests/option_unwrap_fn_test.v
+++ b/vlib/v/tests/option_unwrap_fn_test.v
@@ -1,0 +1,64 @@
+@[heap]
+struct Chan[T] {
+	c chan T
+}
+
+pub type EventListener[T] = fn (T) !
+
+pub type Check[T] = fn (T) bool
+
+struct EventWaiter[T] {
+	check ?Check[T]
+	c     Chan[T]
+}
+
+pub struct EventController[T] {
+mut:
+	id        int
+	wait_fors map[int]EventWaiter[T]
+	listeners map[int]EventListener[T]
+}
+
+fn (mut ec EventController[T]) generate_id() int {
+	return ec.id++
+}
+
+pub fn (ec EventController[T]) emit(e T) bool {
+	for i, w in ec.wait_fors {
+		mut b := false
+		c := w.check
+		if c != none {
+			b = c(e)
+		} else {
+			b = true
+		}
+		println('${i} passed: ${b}')
+		return b
+	}
+	return false
+}
+
+struct Foo {}
+
+struct Bar {}
+
+fn use[T](z EventController[T]) bool {
+	return z.emit(T{})
+}
+
+fn test_main() {
+	assert use(EventController[Foo]{
+		wait_fors: {
+			0: EventWaiter[Foo]{
+				check: fn (t Foo) bool {
+					return false
+				}
+			}
+		}
+	}) == false
+	assert use(EventController[Bar]{
+		wait_fors: {
+			0: EventWaiter[Bar]{}
+		}
+	}) == true
+}

